wsmaster/che-core-api-factory-bitbucket-server/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketUrl.java

```diff
@@ -0,0 +1,190 @@
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.factory.server.bitbucket;
+
+import com.google.common.base.Strings;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.StringJoiner;
+import java.util.stream.Collectors;
+import org.eclipse.che.api.factory.server.urlfactory.RemoteFactoryUrl;
+
+/** Representation of a bitbucket URL, allowing to get details from it. */
+public class BitbucketUrl implements RemoteFactoryUrl {
+
+  /** Hostname of bitbucket URL */
+  private String hostName;
+
+  /** Project part of bitbucket URL */
+  private String project;
+
+  /** Repository part of the URL. */
+  private String repository;
+
+  /** Branch name */
+  private String branch;
+
+  /** Factory json filename */
+  private String factoryFilename;
+
+  /** Devfile filenames list */
+  private final List<String> devfileFilenames = new ArrayList<>();
+
+  /**
+   * Creation of this instance is made by the parser so user may not need to create a new instance
+   * directly
+   */
+  protected BitbucketUrl() {}
+
+  /**
+   * Gets hostname of this bitbucket url
+   *
+   * @return the project part
+   */
+  public String getHostName() {
+    return this.hostName;
+  }
+
+  public BitbucketUrl withHostName(String hostName) {
+    this.hostName = hostName;
+    return this;
+  }
+
+  /**
+   * Gets project of this bitbucket url
+   *
+   * @return the project part
+   */
+  public String getProject() {
+    return this.project;
+  }
+
+  public BitbucketUrl withProject(String project) {
+    this.project = project;
+    return this;
+  }
+
+  /**
+   * Gets repository of this bitbucket url
+   *
+   * @return the repository part
+   */
+  public String getRepository() {
+    return this.repository;
+  }
+
+  protected BitbucketUrl withRepository(String repository) {
+    this.repository = repository;
+    return this;
+  }
+
+  protected BitbucketUrl withDevfileFilenames(List<String> devfileFilenames) {
+    this.devfileFilenames.addAll(devfileFilenames);
+    return this;
+  }
+
+  /**
+   * Gets factory file name of this bitbucket url
+   *
+   * @return the factory file name
+   */
+  @Override
+  public String getFactoryFilename() {
+    return this.factoryFilename;
+  }
+
+  protected BitbucketUrl withFactoryFilename(String factoryFilename) {
+    this.factoryFilename = factoryFilename;
+    return this;
+  }
+
+  /**
+   * Gets branch of this bitbucket url
+   *
+   * @return the branch part
+   */
+  public String getBranch() {
+    return this.branch;
+  }
+
+  protected BitbucketUrl withBranch(String branch) {
+    if (!Strings.isNullOrEmpty(branch)) {
+      this.branch = branch;
+    }
+    return this;
+  }
+
+  /**
+   * Provides location to raw content of the factory json file
```

Comment:
can we omit that?

Reply:
no since it's interface method

Reply:
can we remove it from the interface if it not used?

Reply:
it is not officially deprecated.